app/scripts/lib/extension-update-manager.ts

```diff
@@ -0,0 +1,77 @@
+/**
+ * Manages extension update detection and application.
+ * Provides a clean interface for detecting and applying extension updates
+ * during idle periods to minimize user disruption.
+ */
+
+import browser from 'webextension-polyfill';
+
+export class ExtensionUpdateManager {
+  private updatePending = false;
+
+  private isIdle = false;
+
+  /**
+   * Initializes the update manager by setting up event listeners
+   * for update detection.
+   */
+  public initialize(): void {
+    // Ensure we start in a non-idle state during initialization
+    this.isIdle = false;
+
+    browser.runtime.onUpdateAvailable.addListener(
+      this.handleUpdateAvailable.bind(this),
+    );
+  }
+
+  /**
+   * Sets the idle state of the extension.
+   *
+   * @param idle - Whether the extension is currently idle
+   */
+  public setIdleState(idle: boolean): void {
+    this.isIdle = idle;
+
+    // If we're now idle and there's a pending update, apply it immediately
+    if (idle && this.updatePending) {
+      this.applyPendingUpdateIfNeeded();
+    }
+  }
+
+  /**
+   * Checks if there's a pending update and applies it if needed.
+   * This should be called during idle periods when it's safe to reload
+   * the extension without disrupting user activity.
+   */
+  public applyPendingUpdateIfNeeded(): void {
+    if (this.updatePending && this.isIdle) {
+      try {
+        browser.runtime.reload();
+      } catch (error) {
+        console.error('Failed to reload extension:', error);
+      }
+    }
+  }
+
+  /**
+   * Handles the onUpdateAvailable event by marking an update as pending.
+   * The listener is removed after the first notification to prevent duplicate handlers.
+   * If the extension is already idle, applies the update immediately.
+   *
+   * @private
+   */
+  private handleUpdateAvailable(): void {
```

Comment:
Nit: We don't use `private` anymore, instead using the `#` prefix, which makes the method fully private even at runtime.

This is probably not enforced by the lint rules we use right now, but it will be in later versions of the lint config.

Reply:
Good to know. Defo will update 384b8c1d92ee0f1d900e78022373db0c88f8c2f2